rust-runtime/aws-smithy-types/src/byte_stream/bytestream_util/http_body_0_4_x.rs

```diff
@@ -0,0 +1,414 @@
+                State::Loading(ref mut future) => {
+                    match futures_core::ready!(Pin::new(future).poll(cx)) {
+                        Ok(file) => {
+                            self.state = State::Loaded {
+                                stream: ReaderStream::with_capacity(
+                                    file.take(self.length),
+                                    self.buffer_size,
+                                ),
+                                bytes_left: self.length,
+                            };
+                        }
+                        Err(e) => return Poll::Ready(Some(Err(e.into()))),
+                    };
+                }
+                State::Loaded {
+                    ref mut stream,
+                    ref mut bytes_left,
+                } => {
+                    use futures_core::Stream;
+                    return match futures_core::ready!(Pin::new(stream).poll_next(cx)) {
+                        Some(Ok(bytes)) => {
+                            *bytes_left -= bytes.len() as u64;
+                            Poll::Ready(Some(Ok(bytes)))
+                        }
+                        None => Poll::Ready(None),
+                        Some(Err(e)) => Poll::Ready(Some(Err(e.into()))),
+                    };
+                }
+            };
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        std::task::Poll::Ready(Ok(None))
+    }
+
+    fn is_end_stream(&self) -> bool {
+        match self.state {
+            State::Unloaded(_) | State::Loading(_) => self.length == 0,
+            State::Loaded { bytes_left, .. } => bytes_left == 0,
+        }
+    }
+
+    fn size_hint(&self) -> http_body_0_4::SizeHint {
+        http_body_0_4::SizeHint::with_exact(self.length)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::byte_stream::{ByteStream, FsBuilder, Length};
+    use bytes::Buf;
+    use http_body_0_4::Body;
+    use std::io::Write;
+    use tempfile::NamedTempFile;
+
+    #[tokio::test]
+    async fn path_based_bytestreams_with_builder() {
+        let mut file = NamedTempFile::new().unwrap();
+
+        for i in 0..10000 {
+            writeln!(file, "Brian was here. Briefly. {}", i).unwrap();
+        }
+        let file_length = file
+            .as_file()
+            .metadata()
+            .expect("file metadata is accessible")
+            .len();
+
+        let body = FsBuilder::new()
+            .path(&file)
+            .buffer_size(16384)
+            .length(Length::Exact(file_length))
+            .build()
+            .await
+            .unwrap()
+            .into_inner();
+
+        // assert that the specified length is used as size hint
+        assert_eq!(body.content_length(), Some(file_length));
+
+        let mut body1 = body.try_clone().expect("retryable bodies are cloneable");
```

Comment:
Nit: `body1` existed before this PR, but maybe good to rename it now to something just `body` or `body_04` since this code is under the `body_0_4` module.

Reply:
Yeah, I think it was copied from a another test where the body was cloned several times and each clone had its own number. I've renamed this to body1 and the test still passes.